src/charm.py

```diff
@@ -90,45 +113,76 @@ def __init__(self, *args):
         # Expose the Kiali workload through the service
         self.unit.set_ports(Port("tcp", KIALI_PORT))
 
-    def on_collect_status(self, e: ops.CollectStatusEvent):
-        """Collect and report the statuses of the charm."""
-        # TODO: Extract the generation of these statuses to a helper method, that way other parts of the charm can know
-        #  if the charm is active too.
+        # TODO: Once this exists in grafana
+        self._grafana_metadata = GrafanaMetadataRequirer(...)
+
+        # TODO: Once this exists in Tempo
+        self._tempo_metadata = TempoMetadataRequirer(...)
+        self._tempo_datasource_exchange = DatasourceExchange(
+            charm=self,
+            requirer_endpoint=TEMPO_DATASOURCE_EXCHANGE_RELAION_NAME
+        )
+        self.framework.observe(self.on[TEMPO_DATASOURCE_EXCHANGE_RELAION_NAME].relation_changed, self.reconcile)
+
+    def reconcile(self, _event: ops.ConfigChangedEvent):
+        """Reconcile the entire state of the charm."""
         statuses: List[StatusBase] = []
 
-        if not self._container.can_connect():
-            statuses.append(ops.WaitingStatus("Waiting for the Kiali container to be ready"))
-
-        if not (prometheus_source_available := self._is_prometheus_source_available()):
-            statuses.append(ops.BlockedStatus("Prometheus source is not available"))
-
-        if prometheus_source_available:
-            # Only valid if we have a prometheus source
-            if not _is_kiali_available(self._internal_url + self._prefix):
-                statuses.append(
-                    ops.WaitingStatus(
-                        "Kiali is configured and container is ready, but Kiali's web server is not available"
-                    )
+        try:
+            prometheus_url = self._get_prometheus_source_url()
+        except ConfigurationBlockingError as e:
+            statuses.append(ops.BlockedStatus(str(e)))
+            prometheus_url = None
+        except ConfigurationWaitingError as e:
+            statuses.append(ops.WaitingStatus(str(e)))
+            prometheus_url = None
```

Comment:
These blocks feel verbose, but I like how encapsulated `._get_prometheus_source_url` is.  It'll be easy to test.